### PR TITLE
Fix the implementation of the `Meshes.XYZ` function

### LINKED ISSUE #819: Fix `convert(StructuredGrid, grid)` case
MWE:
```julia
grid = CartesianGrid((0.,0.,0.), (100.,100.,100.), dims=(50, 25, 20))
convert(StructuredGrid, grid)
```

### PR DESCRIPTION
**Pull request by eliascarv**
closes #819

### CODE REVIEW

File: src/mesh/rectilineargrid.jl [hunk 1]
```diff
@@ -35,12 +35,14 @@ vertex(g::RectilinearGrid{Dim}, ijk::Dims{Dim}) where {Dim} = Point(getindex.(g.
...
+XYZ(g::RectilinearGrid{Dim,T}) where {Dim,T} = _genXYZ(T, xyz(g))
+
+@generated function _genXYZ(::Type{T}, xyz::NTuple{Dim}) where {T,Dim}
   exprs = ntuple(Dim) do d
     quote
-      a = g.xyz[$d]
-      N = length(a)
-      A = Array{T,Dim}(undef, @ntuple($Dim, i -> N))
+      dims = @ntuple($Dim, i -> length(xyz[i]))
+      A = Array{T,Dim}(undef, dims)
```
Comment (juliohm) [inline src/mesh/rectilineargrid.jl]:
```suggestion
      A = Array{T,Dim}(undef, length.(xyz))
```

File: src/mesh/rectilineargrid.jl [hunk 2]
```diff
@@ -35,12 +35,13 @@ vertex(g::RectilinearGrid{Dim}, ijk::Dims{Dim}) where {Dim} = Point(getindex.(g.
 
 xyz(g::RectilinearGrid) = g.xyz
 
-@generated function XYZ(g::RectilinearGrid{Dim,T}) where {Dim,T}
+XYZ(g::RectilinearGrid{Dim,T}) where {Dim,T} = xyzXYZ(T, xyz(g))
+
+@generated function xyzXYZ(::Type{T}, xyz::NTuple{Dim}) where {T,Dim}
```
Comment (eliascarv) [inline src/mesh/rectilineargrid.jl]: Maybe this code could be like this:
```suggestion
@generated function xyzXYZ(xyz::NTuple{Dim,<:AbstractVector{T}}) where {Dim,T}
```

File: src/mesh/rectilineargrid.jl [hunk 3]
```diff
@@ -35,12 +35,13 @@ vertex(g::RectilinearGrid{Dim}, ijk::Dims{Dim}) where {Dim} = Point(getindex.(g.
...
   exprs = ntuple(Dim) do d
     quote
-      a = g.xyz[$d]
-      N = length(a)
-      A = Array{T,Dim}(undef, @ntuple($Dim, i -> N))
+      A = Array{T,Dim}(undef, length.(xyz))
+      a = xyz[$d]
```
Comment (eliascarv) [inline src/mesh/rectilineargrid.jl]: This order looks better visually:
```suggestion
      a = xyz[$d]
      A = Array{T,Dim}(undef, length.(xyz))
```

File: src/mesh/cartesiangrid.jl [hunk 4]
```diff
@@ -125,7 +125,7 @@ function xyz(g::CartesianGrid{Dim}) where {Dim}
   end
 end
 
-XYZ(g::CartesianGrid) = XYZ(convert(RectilinearGrid, g))
+XYZ(g::CartesianGrid{Dim,T}) where {Dim,T} = xyzXYZ(T, xyz(g))
```
Comment (eliascarv) [inline src/mesh/cartesiangrid.jl:128]: Fix code:
```suggestion
XYZ(g::CartesianGrid{Dim,T}) where {Dim,T} = xyzXYZ(xyz(g))
```

File: src/mesh/rectilineargrid.jl [hunk 5]
```diff
@@ -35,12 +35,13 @@ vertex(g::RectilinearGrid{Dim}, ijk::Dims{Dim}) where {Dim} = Point(getindex.(g.
 
 xyz(g::RectilinearGrid) = g.xyz
 
-@generated function XYZ(g::RectilinearGrid{Dim,T}) where {Dim,T}
+XYZ(g::RectilinearGrid{Dim,T}) where {Dim,T} = xyzXYZ(T, xyz(g))
```
Comment (eliascarv) [inline src/mesh/rectilineargrid.jl:38]: Fix code:
```suggestion
XYZ(g::RectilinearGrid{Dim,T}) where {Dim,T} = xyzXYZ(xyz(g))
```

File: src/mesh/rectilineargrid.jl [hunk 6]
```diff
@@ -35,12 +35,13 @@ vertex(g::RectilinearGrid{Dim}, ijk::Dims{Dim}) where {Dim} = Point(getindex.(g.
...
+
+@generated function xyzXYZ(xyz::NTuple{Dim,<:AbstractVector{T}}) where {Dim,T}
```
Comment (juliohm) [inline src/mesh/rectilineargrid.jl]: Now that the function accepts a NTuple, we can rename it to `XYZ` and rely on dispatch

Comment (juliohm) [reply src/mesh/rectilineargrid.jl]: That means that we will have XYZ with `Grid` types and with `NTuple` to convert the `xyz` to `XYZ`

File: src/mesh/rectilineargrid.jl [hunk 7]
```diff
@@ -35,12 +35,13 @@ vertex(g::RectilinearGrid{Dim}, ijk::Dims{Dim}) where {Dim} = Point(getindex.(g.
 
 xyz(g::RectilinearGrid) = g.xyz
 
-@generated function XYZ(g::RectilinearGrid{Dim,T}) where {Dim,T}
+XYZ(g::RectilinearGrid) = XYZ(xyz(g))
+
+@generated function XYZ(xyz::NTuple{Dim,<:AbstractVector{T}}) where {Dim,T}
```
Comment (eliascarv) [inline src/mesh/rectilineargrid.jl]: Should we leave this function in this file? Or do we move it to another file?

Comment (juliohm) [reply src/mesh/rectilineargrid.jl]: Let's move it to utils.jl

**Review by juliohm — APPROVED** [review]

**Pull request merged**